### PR TITLE
[Merged by Bors] - feat: relate List.mapAccumr to List.foldr

### PR DESCRIPTION
**Pull request by alexkeizer**
Add lemmas that rewrite an application of `List.mapAccumr` or `List.mapAccumr_2` into an application of `List.foldr`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Lemmas.lean [hunk 1]
```diff
@@ -82,4 +83,48 @@ theorem foldl_range_eq_of_range_eq {f : α → β → α} {g : α → γ → α}
...
+  ### MapAccumr and Foldr
+  Some lemmas relation `mapAccumr` and `foldr`
+-/
+section MapAccumr
+
+theorem mapAccumr_eq_foldr (f : α → σ → σ × β) (as : List α) (s : σ) :
+    mapAccumr f as s = List.foldr (fun a (s, bs) =>
+                                    let ⟨s_out, b⟩ := f a s
+                                    (s_out, b :: bs)
+                                  ) (s, []) as := by
```
Comment (ChrisHughes24) [inline Mathlib/Data/List/Lemmas.lean:98]: I think doing this pattern matching in a theorem statement generally makes proofs harder and maybe it's better to just use the projections instead. I don't have super strong opinions on this, but it's certainly something I've never see in mathlib.

Comment (alexkeizer) [reply Mathlib/Data/List/Lemmas.lean]: Fair point, I've changed it to projections

**Comment by alexkeizer** [comment]
Whoops, I tried to prove the second one with just the simp call as well, but didn't realize a final `rfl` was needed too.

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**